src/classes/actions-cache/upload.ts

```diff
@@ -0,0 +1,180 @@
+  return;
+};
+
+const resetCacheWithOctokit = async (cacheKey: string): Promise<void> => {
+  const token = core.getInput('repo-token');
+  const client = getOctokit(token, undefined, octokitRetry);
+  // TODO: better way to get repository?
+  const repo = process.env['GITHUB_REPOSITORY'];
+  core.debug(`remove cache "${cacheKey}"`);
+  try {
+    // TODO: replace with client.rest.
+    await client.request(
+      `DELETE /repos/${repo}/actions/caches?key=${cacheKey}`
+    );
+  } catch (error) {
+    if (error.status) {
+      core.debug(`Cache ${cacheKey} does not exist`);
+    } else {
+      throw error;
+    }
+  }
+};
+
+const reserveCache = async (
+  httpClient: HttpClient,
+  fileSize: number,
+  cacheKey: string,
+  cacheVersion: string
+): Promise<number> => {
+  const reserveCacheRequest = {
+    key: cacheKey,
+    version: cacheVersion,
+    cacheSize: fileSize
+  };
+  const response = await retryTypedResponse('reserveCache', async () =>
+    httpClient.postJson<{cacheId: number}>(
+      getCacheApiUrl('caches'),
+      reserveCacheRequest
+    )
+  );
+
+  // handle 400 in the special way
+  if (response?.statusCode === 400)
+    throw new Error(
+      response?.error?.message ??
+        `Cache size of ~${Math.round(
+          fileSize / (1024 * 1024)
+        )} MB (${fileSize} B) is over the data cap limit, not saving cache.`
+    );
+
+  const cacheId = response?.result?.cacheId;
+
+  if (cacheId === undefined)
+    throw new ReserveCacheError(
+      `Unable to reserve cache with key ${cacheKey}, another job may be creating this cache. More details: ${response?.error?.message}`
+    );
+  return cacheId;
+};
+
+const commitCache = async (
+  httpClient: HttpClient,
+  cacheId: number,
+  filesize: number
+): Promise<void> => {
+  const response = (await retryTypedResponse('commitCache', async () =>
+    httpClient.postJson<null>(getCacheApiUrl(`caches/${cacheId.toString()}`), {
+      size: filesize
+    })
+  )) as TypedResponse<null>;
+  if (!isSuccessStatusCode(response.statusCode)) {
+    throw new Error(
+      `Cache service responded with ${response.statusCode} during commit cache.`
+    );
+  }
+};
+
+export const uploadFileToActionsCache = async (
+  filePath: string,
+  cacheKey: string,
+  cacheVersion: string
+) => {
+  try {
+    await resetCacheWithOctokit(cacheKey);
+    const fileSize = fs.statSync(filePath).size;
+
+    if (fileSize === 0) {
+      core.info(`the cache ${cacheKey} will be removed`);
+      return;
+    }
+
+    const httpClient = createHttpClient();
+
+    const cacheId = await reserveCache(
+      httpClient,
+      fileSize,
+      cacheKey,
+      cacheVersion
+    );
+
+    await uploadFile(httpClient, cacheId, filePath, fileSize);
+
+    await commitCache(httpClient, cacheId, fileSize);
+  } catch (error) {
+    const typedError = error as Error;
+    if (typedError.name === ValidationError.name) {
+      throw error;
+    } else if (typedError.name === ReserveCacheError.name) {
```

Comment:
`else` is pointless here as early return (in form of throw) is implemented in first clause. Perhaps we can take advantage of early return further down the chain as well.

Reply:
done